activation_key_logger.go

```diff
@@ -0,0 +1,190 @@
+	levelError
+)
+
+var (
+	levelMapping = map[string]levelID{
+		"debug":   levelDebug,
+		"info":    levelInfo,
+		"warning": levelWarning,
+		"error":   levelError,
+	}
+)
+
+type levelID byte
+
+type ActivationLoggerConfig struct {
+	Underlying Logger
+
+	Activations map[string]interface{}
+}
+
+type activationLogger struct {
+	underlying Logger
+
+	activations map[string]interface{}
+}
+
+func NewActivation(config ActivationLoggerConfig) (Logger, error) {
+	if config.Underlying == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Underlying must not be empty", config)
+	}
+
+	l := &activationLogger{
+		underlying: config.Underlying,
+
+		activations: config.Activations,
+	}
+
+	return l, nil
+}
+
+func (l *activationLogger) Log(keyVals ...interface{}) error {
+	activated, err := shouldActivate(l.activations, keyVals)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if activated {
+		return l.underlying.Log(keyVals...)
+	}
+
+	return nil
+}
+
+func (l *activationLogger) LogCtx(ctx context.Context, keyVals ...interface{}) error {
+	activated, err := shouldActivate(l.activations, keyVals)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if activated {
+		return l.underlying.LogCtx(ctx, keyVals...)
+	}
+
+	return nil
+}
+
+func (l *activationLogger) With(keyVals ...interface{}) Logger {
+	return l.underlying.With(keyVals...)
+}
+
+func containsKey(keyVals []interface{}, aKey string) bool {
+	for i := 0; i < len(keyVals); i += 2 {
+		s, ok := keyVals[i].(string)
+		if ok && s == aKey {
+			return true
+		}
+	}
+
+	return false
+}
+
+func containsVal(keyVals []interface{}, aVal interface{}) bool {
+	for i := 1; i < len(keyVals); i += 2 {
+		if keyVals[i] == aVal {
+			return true
+		}
+	}
+
+	return false
+}
+
+func isLevelAllowed(keyVals []interface{}, aVal interface{}) bool {
+	s, ok := aVal.(string)
+	if !ok {
+		return false
+	}
+	activationLevel, ok := levelMapping[s]
+	if !ok {
+		return false
+	}
+
+	for i := 0; i < len(keyVals); i += 2 {
+		k, ok := keyVals[i].(string)
+		if !ok {
+			continue
+		}
+		if k != KeyLevel {
+			continue
+		}
+		v, ok := keyVals[i+1].(string)
+		if !ok {
+			continue
+		}
+		keyValsLevel, ok := levelMapping[v]
+		if !ok {
+			continue
+		}
+
+		return activationLevel >= keyValsLevel
+	}
+
+	return false
+}
+
+func isVerbosityAllowed(keyVals []interface{}, aVal interface{}) bool {
+	activationVerbosity, ok := aVal.(int)
+	if !ok {
+		return false
+	}
+
+	for i := 0; i < len(keyVals); i += 2 {
+		k, ok := keyVals[i].(string)
+		if !ok {
+			continue
+		}
+		if k != KeyVerbosity {
+			continue
+		}
+		keyValsVerbosity, ok := keyVals[i+1].(int)
+		if !ok {
+			continue
+		}
+
+		return activationVerbosity >= keyValsVerbosity
+	}
+
+	return false
+}
+
+func shouldActivate(activations map[string]interface{}, keyVals []interface{}) (bool, error) {
+	var activationCount int
+
+	for aKey, aVal := range activations {
+		if containsKey(keyVals, aKey) && containsVal(keyVals, aVal) {
```

Comment:
Note that activation key matching is not limited to special implementations like the reserved level and verbosity keys but can also be any kind of arbitrary key-value pairs.

Reply:
This is not an objection or criticism by itself, but a thought that came to my mind when going through this:
Can there be any risk of accidental activation here? Of course it would always be on the programmer writing the logging line, but having more than one maintainer in a project there could be possibility of confusion.

One thing is also that I think this logger shouldn't be used in performance sensitive parts because linear search for matching key & matching value in a loop during one logging function call can bring huge overhead when used in wrong place. Maybe take this into consideration when writing godoc for this?

Reply:
Any objections and criticisms are welcome. No worries. 

I think in rare cases there can be confusion and something will be logged that was not intended to be logged. I don't think something can be hidden by default just because somebody uses these features wrong. I think the potential is quite low though but maybe you have a concrete example in mind. Then please share. 

Performance is for sure important in certain cases. For instance our microservices and operators should care about these things. I do not intent to use these features here in any daemons though. There we should just print everything all the time and the logging infrastructure provides tools to analyse and filter structure data provided by microservices. So this here as described in the PR description is for command line tools like `opsctl` and I do not think that the impacts of the algorithms here are critical for such applications.

Reply:
Didn't have anything specific in mind, but an example could be perhaps a case where one developer enables tuple `("info", "pod")` and later another developer enables `("debug", "deployment")` and commits it after which the first developer might be wondering where the messages came from, but I think this is pretty far fetched, especially as you mentioned that this is intended for CLI tools.

All good here. :)